### ISSUE TITLE
[request] Separate TailwindCSS bundle for each entry point

### ISSUE THREAD

**Issue by cawa-93** [roots/bud#2537]
### Agreement

- [X] This is not a duplicate of an existing issue
- [X] I have read the [guidelines for Contributing to Roots Projects](https://github.com/roots/.github/blob/master/CONTRIBUTING.md)
- [X] This is not a personal support request that should be posted on the [Roots Discourse](https://discourse.roots.io/) community

### The feature

At the moment, tailwindcss generates classes based on the template of the WHOLE site. This is not very efficient considering that WordPress is a multi-page system with an explicit template hierarchy. 

### Use case

In Sage theme I have 

```blade
index.blade.php

<div class="bg-red-500"></div>
<div class="bg-blue-500"></div>
```

```blade
category.blade.php

<div class="bg-red-500"></div>
<div class="bg-green-500"></div>
```

Currently, tailwindcss will generate single CSS based on both templates:
```css
.bg-red-500 { ... }
.bg-blue-500 { ... }
.bg-green-500 { ... }
```
That means every page will contain at least one unused CSS class definition. 

For better performance I would like to get separated CSS files for each WordPress template type that containe only CSS that actually used in it. 
```css
/* index.css */
.bg-red-500 { ... }
.bg-blue-500 { ... }
```
```css
/* category.css */
.bg-red-500 { ... }
.bg-green-500 { ... }
```

**Comment by cawa-93** [comment]
As PoC, I wrote something like this:
Separate tailwind config for each entry point. In each config, it parses file tree and resolve views and components used for the entry. For example, for category page:

```js
// ./config/tailwind.category.config.js

// extend tailwind config from project root, but overwrite content
import baseConfig from '../tailwind.config.js';

export default {
    ...baseConfig,
    content: findDeps('category.blade.php'),
};
```

<details><summary>`findDeps` implementation</summary>

```js
import fs from 'fs';

const hashMap = new Map();

export function findDeps(entry, viewsRoot = './resources/views/', appRoot = './app/View/Components/') {
    return [...getFiles({file: viewsRoot + entry, visited: new Set})];

    /**
     * @param {string} file
     * @param {Set<string>} visited
     * @return {Set<string>}
     */
    function getFiles({file, visited}) {

        if (hashMap.has(file)) {
            console.log('[CACHE]', file);
            return hashMap.get(file);
        }


        if (visited.has(file)) {
            console.log('[VISITED]', file);
            return visited;
        }

        visited.add(file);

        if (!fs.existsSync(file)) {
            console.log('[NOT EXIST]', file);
            return visited;
        }


        console.log('[NEW]', file);

        const content = fs.readFileSync(file, 'utf8');

        const deps = file.endsWith('.blade.php')
            ? resolveDepsBlade({content, visited})
            : resolveViewFromComponentClass({content, visited});

        hashMap.set(file, deps);

        return deps;
    }

    /**
     * @param {string} content
     * @param {Set<string>} visited
     * @return {Set<string>}
     */
    function resolveDepsBlade({content, visited}) {
        visited = resolveLayout({content, visited});
        visited = resolveComponentsFromBlade({content, visited});
        return visited;
    }

    /**
     * @param {string} content
     * @param {Set<string>} visited
     * @return {Set<string>}
     */
    function resolveLayout({content, visited}) {
        const layouts = content.match(/(?<=@(extends|include)\(\s*')[a-z0-9._-]+(?='\s*\))/ig);

        if (!layouts) {
            return visited;
        }

        layouts.map(s => viewsRoot + s.replaceAll('.', '/') + '.blade.php').forEach(file => {
            visited = getFiles({file, visited});
        });
        return visited;
    }

    /**
     * @param {string} content
     * @param {Set<string>} visited
     * @return {Set<string>}
     */
    function resolveComponentsFromBlade({content, visited}) {
        const components = content.match(/(?<=<\s*x-)[a-z0-9._-]+/ig);

        if (!components) {
            return visited;
        }

        for (const component of components) {
            const filePathChunks = component.split('.').map(toPascalCase);
            const file = appRoot + filePathChunks.join('/') + '.php';
            visited = getFiles({file, visited});
        }

        return visited;
    }

    function resolveViewFromComponentClass({content, visited}) {
        const views = content.match(/components\.[a-z0-9._-]+/ig);
        if (!views) {
            return visited;
        }


        views.map(s => viewsRoot + s.replaceAll('.', '/') + '.blade.php').forEach(file => {
            visited = getFiles({file, visited});
        });
        return visited;
    }
}

function toPascalCase(text) {
    return text.replace(/(^\w|-\w)/g, clearAndUpper);
}

function clearAndUpper(text) {
    return text.replace(/-/, '').toUpperCase();
}

```
</details>

If `views/category.blade.php` has markup:
```blade
@extends('layouts.app')

@section('content')
    <x-load-more-post />
@endsection
```
Then `content` will be 
```js
[
  'resources/views/category.blade.php', 
  'resources/views/layouts/app.blade.php', 
  'app/Views/Components/LoadMorePost.php',
  'resources/views/components/load-more-post.blade.php',
]
```

Then create `category.css` and refer to this config:
```css
@config "./config/tailwind.category.config.js";

@tailwind base;
@tailwind components;
@tailwind utilities;
```

Now category.css will contain only that tailwind styles what was used in category page directly or in child components of any depth.

**Comment by kellymears** [comment]
I think this is best handled with a multi-compiler configuration.

Basic setup for Sage involves changes to three files: 

1. setting up multi-compiler in `bud.config.js`
2. overriding `config/assets.php`
3. modifying `app/setup.php`

I'll give an example where we treat app and editor as separate compilations. You can add as many of these as you need.

## 1. bud.config.js

Use the outermost scope to configure the dev server, and use callbacks to configure each instance:

```js
export default async (bud) => {
  bud
    .setUrl(`https://flywheel.local:3000`)
    .setProxyUrl(`https://flywheel.local`)
    .watch(['resources/views', 'app']);

  await bud.make(`app`, async (app) => {
    app
      .setPath(`@dist`, `public/app`)
      .setPublicPath('/wp-content/themes/sage/public/app/')
      .entry('main', ['@scripts/app', '@styles/app']);
  });

  await bud.make(`editor`, async (editor) => {
    editor
      .setPath(`@dist`, `public/editor`)
      .setPublicPath('/wp-content/themes/sage/public/editor/')
      .entry('main', ['@scripts/editor', '@styles/editor']);
  });
};
```

## 2. config/assets.php

Tell acorn about the structure of our assets:

```php
<?php

return [
    'default' => 'app',

    'manifests' => [
        'app' => [
            'path' => get_theme_file_path('public/app'),
            'url' => get_theme_file_uri('public/app'),
            'assets' => get_theme_file_path('public/app/manifest.json'),
            'bundles' => get_theme_file_path('public/app/entrypoints.json'),
        ],

        'editor' => [
            'path' => get_theme_file_path('public/editor'),
            'url' => get_theme_file_uri('public/editor'),
            'assets' => get_theme_file_path('public/editor/manifest.json'),
            'bundles' => get_theme_file_path('public/editor/entrypoints.json'),
        ]
    ]
];
```

## 3. app/setup.php

Hook Acorn to WordPress:

```php
<?php

namespace App;

use function Roots\bundle;

add_action('wp_enqueue_scripts', function () {
    bundle('app', 'main')->enqueue();
}, 100);

add_action('enqueue_block_editor_assets', function () {
    bundle('editor', 'main')->enqueue();
}, 100);

// ...
```

This setup is super flexible and processed in parallel. There is really very minimal overhead:

```sh
╭ app [1/2] [6abc738cf1d04d0d] ./public/app
│
│ main
│  ◉ js/runtime.9c60ae.js ✔ 1.04 kB
│  ◉ css/main.f77734.css  ✔ 5.76 kB
│  ◉ js/main.35ee80.js    ✔ 394 bytes
│
╰ 141ms 8 modules [8/8 modules cached]

╭ editor [2/2] [f70a2e1d844ae988] ./public/editor
│
│ main
│  ◉ js/runtime.97a493.js   ✔ 1.33 kB
│  ◉ css/main.ef46db.css    ✔ ø
│  ◉ js/main.2adfe2.js      ✔ 1.8 kB
│
╰ 140ms 16 modules [16/16 modules cached]

✨  Done in 1.27s.
```

I frequently include bud.js in bedrock and build all plugins and themes from one bud config using this method. The largest project I have consulted on has a dozen themes and numerous plugins built from a single `bud build` command. So, it should work for what you want to do but also scales well for more complex setups.

**Comment by cawa-93** [comment]
I don't really understand how in your example I can tell tailwind which files it should analyze to generate a css for each bundle

**Comment by kellymears** [comment]
First, sorry my answer didn't help with your question. I've reread it now and think I understand better what you are asking. 

The answer is largely the same -- you will set up multiple compilers and configure each instance to your needs.

So, for example, if I wanted to have a different tailwind configuration applied to my page template vs. my search template:

```js
export default async (bud) => {
  bud
    .setUrl(`http://example.test:3000`)
    .setProxyUrl(`http://example.test`)
    .watch(['resources/views', 'app']);

  await bud.make(`page`, async (page) => {
    page
      .setPath(`@dist`, `public/page`)
      .setPublicPath('/wp-content/themes/sage/public/page/')
      .entry('main', ['@styles/page'])

      .tailwind.setConfig((config) => ({
        ...config,
        content: ['resources/views/page.blade.php'],
      }));
  });

  await bud.make(`search`, async (search) => {
    search
      .setPath(`@dist`, `public/search`)
      .setPublicPath('/wp-content/themes/sage/public/search/')
      .entry('main', ['@styles/search'])

      .tailwind.setConfig((config) => ({
        ...config,
        content: ['resources/views/search.blade.php'],
      }));
  });
};
```

**Comment by kellymears** [comment]
I think anything beyond what we're providing with this API would probably be best addressed with a tailwindcss feature request. 

I want bud.js to work well with tailwindcss and provide a few quality of life extras (like customizing a shared tailwind config in different compiler instances, as demonstrated), but I don't want to outright be adding features to tailwindcss. And, I think that's what we'd be doing if your API were implemented.

**Comment by cawa-93** [comment]
Thank you very much. I understand now. I had lost the fact that Bud can change the config in this way. Thanks again. 

Initially, I didn't think about adding a tailwindcss feature. I was thinking more about a sage-specific(?) feature or a wp-specific feature. Something like a blade parser to statically parse template files and create a map. Something like: 
```js
{
  templateDependencies: {
    page: ['resources/views/page.blage.php', 'app/Components/NavMenu.php'],
    search: ['resources/views/search.blage.php', 'app/Components/NavMenu.php'],
    // ...
  }
}
```
So then devs can use it in tailwindcss, unocss, purgecss or any other tools that rely on HTML parsing.
